### PR TITLE
Create HTTP ingestion engine

### LINKED ISSUE #73: Create HTTP ingestion engine
#### Description
We need an HDHomeRun ingestion engine, which streams MPEG-TS over HTTP. Therefore, it should make sense to write a generic HTTP ingestion engine first that the HDHomeRun ingestion engine can inherit from and decorate (see [this comment](https://github.com/tvkitchen/tv-kitchen/issues/68#issuecomment-650600592)). This way we write the most minimal engine possible, test HDHomeRun functionality before enhancing it with HDHomeRun-specifics and accommodate  other HTTP streaming sources (really any TS file on a server).

#### Relevant Resources / Research
- [HDHomeRun HTTP Development Guide](https://www.silicondust.com/hdhomerun/hdhomerun_http_development.pdf)

#### Related Issues
#68

### PR DESCRIPTION
**Pull request by chriszs**
## Description
Creates a HTTP ingestion engine that can serve as a base for a HDHomeRun engine.

## Due Diligence Checklist
- [x] Tests
- [x] Consolidated commits
- [x] Relevant labels assigned

## Steps to Test
1. `yarn test`

## Related Issues
Resolves #73

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -42,7 +42,8 @@
     "@tvkitchen/base-constants": "^1.0.0",
     "@tvkitchen/base-classes": "^1.1.0",
     "@tvkitchen/base-errors": "^1.0.0",
-    "@jest/console": "^25.1.0"
+    "@jest/console": "^25.1.0",
+    "node-fetch": "^2.6.0"
```
Comment (slifty) [inline package.json:46]: We've been tending to put new dependencies in their own commit (e.g. "Add node-fetch package") and I think that would be good to continue in this PR.

I appreciate that you explain a bit in this commit as to why we're using node-fetch vs something native in node but a quick extra note would be helpful (like, one sentence; not re-producing an entire "fetch vs http" blog post or whatnot)!

File: src/components/ingestion/HttpIngestionEngine.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+	 *
+	 * @param  {string}  url The url of the stream to be ingested
+	 */
+	constructor(url) {
+		if (!url) {
+			throw new Error(
+				'HttpIngestionEngine must be instantiated with a stream URL',
+			)
+		}
+		super()
```
Comment (slifty) [inline src/components/ingestion/HttpIngestionEngine.js:28]: In Java I believe you *must* call super first; I appreciate that we don't need to reference `this` in our validation here, and so validation first works, but in the general case it's *possible* a validation rule might rely on the property of an inherited class and for that reason maybe standardizing around calling `super()` right away is worthwhile.

(if you disagree, plz disregard this comment!)

File: package.json (hunk 1 above)
Comment (slifty) [reply package.json:46]: I know tweaking history is not a common pattern so here's the steps in case they're handy:

* `git rebase -i`
* mark this commit as "edit"
* create a *NEW* commit (while rebasing) that contains just the package + yarn.lock
* continue the rebase with the rest of the commits

File: src/components/ingestion/HttpIngestionEngine.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+		this.url = url
+	}
+
+	/**
```
Comment (slifty) [inline src/components/ingestion/HttpIngestionEngine.js:33]: Since getInputStream is an override method, rather than replicate the documentation lets use the [`@inheritdoc`](https://jsdoc.app/tags-inheritdoc.html) tag.

NOTE: I think I need to make an issue about this since we're totally not doing that elsewhere.

File: src/components/ingestion/__test__/HttpIngestionEngine.unit.test.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+			})
+
+			const fileEngine = new HttpIngestionEngine(
+				url,
+			)
+
+			const inputStream = fileEngine.getInputStream()
+
+			inputStream.on('error', (err) => {
+				expect(err.message).toEqual(`request to ${url} failed, reason: ${message}`)
```
Comment (slifty) [inline src/components/ingestion/__test__/HttpIngestionEngine.unit.test.js:44]: I know the drama that led to checking messages here, but I have concern that the content of an error message is too rigid for a test (e.g. the spec is not that the message looks like this exactly, but rather that the error is of the expected type)

File: src/components/ingestion/__test__/HttpIngestionEngine.unit.test.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+				done()
```
Comment (slifty) [inline src/components/ingestion/__test__/HttpIngestionEngine.unit.test.js:45]: If no error is sent this test will fail but it will fail based on timeout settings, right?  I would prefer a more explicit test that will fail directly if no error has been thrown (who knows if we will have to play with test timeout settings in some environment some day).

Would it be reasonable to modify this test so the `error` event sets a value and the `finish` event is the place that actually runs the `expect` statements in terms of those values?  For instance, having `error` append the error to an `errorCache` array, and then in `finish` have something like:

`expect(errorCache.length).toBe(1)`

Comment (chriszs) [reply src/components/ingestion/__test__/HttpIngestionEngine.unit.test.js:45]: I looked into this, but problem I believe I was having that it didn't emit finish. My research at the time indicated that the error event ended the stream, emitting nothing else, and this was expected and normal. But I'll check again.

Comment (chriszs) [reply src/components/ingestion/__test__/HttpIngestionEngine.unit.test.js]: Okay, used Jest's promise resolution/rejection support for this. Seems better.

File: src/components/ingestion/__test__/HttpIngestionEngine.unit.test.js (hunk 4 above)
Comment (chriszs) [reply src/components/ingestion/__test__/HttpIngestionEngine.unit.test.js]: Changed.

File: src/components/ingestion/HttpIngestionEngine.js (hunk 3 above)
Comment (chriszs) [reply src/components/ingestion/HttpIngestionEngine.js:33]: Okay.

File: src/components/ingestion/HttpIngestionEngine.js (hunk 2 above)
Comment (chriszs) [reply src/components/ingestion/HttpIngestionEngine.js]: Changed.

File: package.json (hunk 1 above)
Comment (chriszs) [reply package.json:46]: Okay.
